### PR TITLE
build: optimize .h generation in GenerateHeaderFrom{Raw,Json}.cmake

### PR DESCRIPTION
**Pull request by l0rinc**
Follow up of the https://github.com/bitcoin/bitcoin/pull/30883 revert.

Replaced multiple file writes with a single string template write.
The raw content is first grouped into 8 byte chunks, followed by another regex replace which wraps them in `std::byte` or just the raw bytes, prefixed with `0x`.

Tested the output with `diff -w` and they're the same - only whitespace differences because slightly different source formatting.

----

Tested the `Raw` performance with:
```bash
time cmake -DRAW_SOURCE_PATH=src/bench/data/block413567.raw -DHEADER_PATH=build/after/block413567.raw.h -DRAW_NAMESPACE=benchmark::data -P cmake/script/GenerateHeaderFromRaw.cmake
```

Before:
> 15.41s user 23.06s system 97% cpu 39.593 total

After:
> 0.77s user 0.06s system 97% cpu 0.849 total

----

Tested the `Json` performance with:
```bash
time cmake -DJSON_SOURCE_PATH=src/secp256k1/src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.json -DHEADER_PATH=build/after/ecdsa_secp256k1_sha256_bitcoin_test.json -P cmake/script/GenerateHeaderFromJson.cmake
````

Before:
> 3.57s user 6.01s system 94% cpu 10.136 total

After:
> 0.17s user 0.01s system 98% cpu 0.187 total

### CODE REVIEW

**Review by maflcko — APPROVED** [review]
Concept ACK, but it would be more consistent and easier to maintain to also transform the json one.

Seems a bit absurd that the regex replacment is faster than appending to a string, so it would be good to test this on Windows and Linux with the minimum and maximum supported cmake version to make sure this is really the last time this needs to be touched.

**Comment by l0rinc** [comment]
> Seems a bit absurd that the regex replacment is faster than appending to a string

I did some simple profiling and this regex version likely works with some mutable stringbuilder in the background, while the previous string solution was likely quadratic since it's immutable, so it was recreated after each byte.
And the file write is likely faster than that since it uses some buffering.

> also transform the json one

I'm still working on that one, but given the concept ACK, I'll do that here as well.

**Comment by hebasto** [comment]
Concept ACK on increasing performance.

File: cmake/script/GenerateHeaderFromRaw.cmake [hunk 1]
```diff
@@ -2,26 +2,26 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or https://opensource.org/license/mit/.
 
+# Extract the base name of the source file
```
[the remark this answers is not in the document]
Comment (l0rinc) [reply cmake/script/GenerateHeaderFromRaw.cmake]: Done

File: cmake/script/GenerateHeaderFromRaw.cmake [hunk 2]
```diff
@@ -2,26 +2,26 @@
...
+get_filename_component(raw_source_basename ${RAW_SOURCE_PATH} NAME_WE)
+
+# Read the raw file content in hexadecimal form
```
[the remark this answers is not in the document]
Comment (l0rinc) [reply cmake/script/GenerateHeaderFromRaw.cmake]: Done

**Comment by l0rinc** [comment]
> it would be good to test this on Windows and Linux with the minimum and maximum supported cmake version to make sure this is really the last time this needs to be touched.

While the speed difference is algorithmic, shouldn't be OS dependent, I would appreciate if someone would do that.

> would be more consistent and easier to maintain to also transform the json one

pushed

**Review by maflcko — APPROVED** [review]
lgtm

File: cmake/script/GenerateHeaderFromJson.cmake [hunk 3]
```diff
@@ -2,25 +2,23 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or https://opensource.org/license/mit/.
 
+get_filename_component(json_source_basename ${JSON_SOURCE_PATH} NAME_WE)
+
 file(READ ${JSON_SOURCE_PATH} hex_content HEX)
-string(REGEX MATCHALL "([A-Za-z0-9][A-Za-z0-9])" bytes "${hex_content}")
+# Split lines every 16 characters and format the bytes into std::byte format
```
[the remark this answers is not in the document]
Comment (l0rinc) [reply cmake/script/GenerateHeaderFromJson.cmake]: removed them

File: cmake/script/GenerateHeaderFromJson.cmake [hunk 4]
```diff
@@ -2,25 +2,23 @@
...
+string(REGEX REPLACE "................" "\\0\n" formatted_bytes "${hex_content}")
+string(REGEX REPLACE "[^\n][^\n]" "0x\\0, " formatted_bytes "${formatted_bytes}")
 
-file(WRITE ${HEADER_PATH} "#include <string_view>\n")
-file(APPEND ${HEADER_PATH} "namespace json_tests{\n")
-get_filename_component(json_source_basename ${JSON_SOURCE_PATH} NAME_WE)
-file(APPEND ${HEADER_PATH} "inline constexpr char detail_${json_source_basename}_bytes[]{\n")
+# Write the content of the .h file
```
[the remark this answers is not in the document]
Comment (l0rinc) [reply cmake/script/GenerateHeaderFromJson.cmake]: removed

File: cmake/script/GenerateHeaderFromJson.cmake [hunk 5]
```diff
@@ -2,25 +2,21 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or https://opensource.org/license/mit/.
 
+get_filename_component(json_source_basename ${JSON_SOURCE_PATH} NAME_WE)
```
Comment (hebasto) [inline cmake/script/GenerateHeaderFromJson.cmake:5]: nit: Can be replaced with the `cmake_path()` command.

From CMake [docs](https://cmake.org/cmake/help/latest/command/get_filename_component.html):
> _Changed in version 3.20:_ This command has been superseded by the [`cmake_path()`](https://cmake.org/cmake/help/latest/command/cmake_path.html#command:cmake_path) command...

File: cmake/script/GenerateHeaderFromJson.cmake [hunk 6]
```diff
@@ -2,25 +2,21 @@
...
+
 file(READ ${JSON_SOURCE_PATH} hex_content HEX)
-string(REGEX MATCHALL "([A-Za-z0-9][A-Za-z0-9])" bytes "${hex_content}")
+string(REGEX REPLACE "................" "\\0\n" formatted_bytes "${hex_content}")
```
Comment (hebasto) [inline cmake/script/GenerateHeaderFromJson.cmake:8]: nit: Using the `string(REPEAT ...)` command to create a string of 16 dots can improve the code's clarity, I think.

**Review by hebasto — COMMENTED** [review]
Changes look correct. Going to benchmark them on Windows tomorrow.

File: cmake/script/GenerateHeaderFromJson.cmake (hunk 6 above)
Comment (l0rinc) [reply cmake/script/GenerateHeaderFromJson.cmake:8]: I thought of that but couldn't find a way to do it inline - and defing a new variable in a separate line just for this makes it less readable
```cmake
string(REPEAT "." 16 line_width)
string(REGEX REPLACE "${line_width}" "\\0\n" formatted_bytes "${hex_content}")
```

File: cmake/script/GenerateHeaderFromJson.cmake (hunk 5 above)
Comment (l0rinc) [reply cmake/script/GenerateHeaderFromJson.cmake:5]: Done, thanks

**Review by hebasto — COMMENTED** [review]
Here are benchmarks on Windows using the following command in PowerShell:
```PowerShell
> (Measure-Command { cmake -DRAW_SOURCE_PATH="src/bench/data/block413567.raw" -DHEADER_PATH="build/bench/data/block413567.raw.h" -DRAW_NAMESPACE="benchmark::data" -P cmake/script/GenerateHeaderFromRaw.cmake }).TotalSeconds
```

- the master branch @ 07c7c96022dd325be1cd3b353d575eb6a5593f55 using `file(APPEND ...)` calls:
```
296.326647
```

- this PR (three consecutive runs)
```
1.6235875
1.5456055
1.5283277
```

Tested on Windows 11 Pro + Visual Studio 2022 17.11.2 with builtin CMake 3.29.5-msvc4.

**Comment by fanquake** [comment]
This needs benchmarking on the systems used by devs (i.e Linux, and ideally those that previously reported issues), to ensure that we don't just have the same problem of Linux performance regressions again.

**Comment by fanquake** [comment]
> I've put the before/after measurements in the commit messages, I'd prefer them separate

You can put that in the PR description, that will be included in the merge.

**Review by hebasto — COMMENTED** [review]
The minimum supported CMake version, 3.22, and the most recent version, 3.30, show similar performance on Linux as in the PR description.

File: cmake/script/GenerateHeaderFromRaw.cmake [hunk 7]
```diff
@@ -2,26 +2,22 @@
...
-endforeach()
+namespace ${RAW_NAMESPACE} {
+inline constexpr std::byte detail_${raw_source_basename}_raw[] {
+${formatted_bytes}
+};
 
-file(APPEND ${HEADER_PATH} "\n};\n")
-file(APPEND ${HEADER_PATH} "inline constexpr std::span ${raw_source_basename}{detail_${raw_source_basename}_raw};\n")
-file(APPEND ${HEADER_PATH} "}")
+inline constexpr std::span ${raw_source_basename}{detail_${raw_source_basename}_raw};
```
Comment (l0rinc) [inline cmake/script/GenerateHeaderFromRaw.cmake:20]: would it make sense to change the return type to `std::span<const std::byte>` instead?

**Comment by l0rinc** [comment]
Thanks a lot for checking @hebasto!

**Review by hebasto — APPROVED** [review]
ACK 2a581144f28bad44de40122864f2f7b9fc5000de.

The new scripts introduce trailing spaces in the generated header, but these are not essential.

**Review by willcl-ark — APPROVED** [review]
tACK 2a581144f28bad44de40122864f2f7b9fc5000de

I didn't review the code in too much detail, but I like hebasto also saw a good speedup on Ubuntu 23.10 x86_64 from 14 seconds down to half a second for block413567.raw with `cmake` 3.30:

```log
will@ubuntu in ~/src/bitcoin on  master [$?⇕] via △ v3.30.3 : 🐍 (bitcoin) took 12s
$ rm -i -Rf build

will@ubuntu in ~/src/bitcoin on  master [$?⇕] via △ v3.30.3 : 🐍 (bitcoin)
$ time cmake -DRAW_SOURCE_PATH=src/bench/data/block413567.raw -DHEADER_PATH=build/before/block413567.raw.h -DRAW_NAMESPACE=benchmark::data -P cmake/script/GenerateHeaderFromRaw.cmake

________________________________________________________
Executed in   14.37 secs    fish           external
   usr time   10.11 secs  936.00 micros   10.11 secs
   sys time    4.13 secs  176.00 micros    4.13 secs


will@ubuntu in ~/src/bitcoin on  master [$?⇕] via △ v3.30.3 : 🐍 (bitcoin) took 14s
$ rm -i -Rf build

will@ubuntu in ~/src/bitcoin on  master [$?⇕] via △ v3.30.3 : 🐍 (bitcoin)
✗ gh pr checkout 30888
From https://github.com/bitcoin/bitcoin
 * [new ref]                 refs/pull/30888/head -> lorinc/cmake-header-optimization
Switched to branch 'lorinc/cmake-header-optimization'

will@ubuntu in ~/src/bitcoin on  lorinc/cmake-header-optimization [$?] via △ v3.30.3 : 🐍 (bitcoin)
$ time cmake -DRAW_SOURCE_PATH=src/bench/data/block413567.raw -DHEADER_PATH=build/after/block413567.raw.h -DRAW_NAMESPACE=benchmark::data -P cmake/script/GenerateHeaderFromRaw.cmake

________________________________________________________
Executed in  673.32 millis    fish           external
   usr time  564.70 millis  851.00 micros  563.85 millis
   sys time  108.13 millis  163.00 micros  107.97 millis
```

The generated headers looked ok, minus whitespace and no newline at end of file.

Overall a nice speedup!

**Pull request merged**